### PR TITLE
ddl: add `addingDefinition` field for adding partition replica check (#18495)

### LINKED ISSUE #18496: DDL: the query to the new added partition which is in it's replica process
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1: make a table with tiflash replica availble
2: add the new table partition to it
3: query to the new added partition

### 2. What did you expect to see? (Required)
the finish of adding partiton DDL should take tiflash replica check into consideration.

### 3. What did you see instead (Required)
None

### 5. Root Cause Analysis
the finish of adding partiton DDL doesn't take tiflash replica check into consideration.

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #18495 to release-4.0

---

Signed-off-by: AilinKid <314806019@qq.com>

### What problem does this PR solve?

Issue Number: close #18496 

Problem Summary: 
It is reported that if a table is in tiflash replica mode and it's available state is `true`, the new added partition will cause query block because the new added partition hasn't been replicated completely.

That's why we wanna add a mid state for adding partiton. In the delete-only state, the `DDL` owner will check for the new added partition's replica info. Once it is completed, then make it public.

### What is changed and how it works?

What's Changed:
1:  add a mid state (delete only for add partiton)
2:  add the rollback logic for add partition 

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Breaking backward compatibility

### Release note 

- add delete only state for add partition replica check
- this pr will cause compatibility problem, when do DDL action in rolling upgrade

### CODE REVIEW

File: ddl/ddl_worker_test.go [hunk 1]
```diff
@@ -862,6 +884,148 @@ func (s *testDDLSuite) TestCancelJob(c *C) {
 	testDropIndex(c, ctx, d, dbInfo, tblInfo, idxOrigName)
 	c.Check(errors.ErrorStack(checkErr), Equals, "")
 	s.checkDropIdx(c, d, dbInfo.ID, tblInfo.ID, idxOrigName, true)
+<<<<<<< HEAD
+=======
```
Comment (zimulala) [inline ddl/ddl_worker_test.go:888]: @AilinKid 
Please fix the conflict.

File: ddl/ddl_worker_test.go [hunk 2]
```diff
@@ -460,6 +460,28 @@ func buildCancelJobTests(firstID int64) []testCancelJob {
 		{act: model.ActionAddPrimaryKey, jobIDs: []int64{firstID + 35}, cancelRetErrs: []error{admin.ErrCancelFinishedDDLJob.GenWithStackByArgs(firstID + 35)}, cancelState: model.StatePublic},
 		{act: model.ActionDropPrimaryKey, jobIDs: []int64{firstID + 36}, cancelRetErrs: noErrs, cancelState: model.StateWriteOnly},
 		{act: model.ActionDropPrimaryKey, jobIDs: []int64{firstID + 37}, cancelRetErrs: []error{admin.ErrCannotCancelDDLJob.GenWithStackByArgs(firstID + 37)}, cancelState: model.StateDeleteOnly},
+<<<<<<< HEAD
+=======
```
Comment (djshow832) [inline ddl/ddl_worker_test.go:464]: It has been many days. Please resolve conflicts.

**Pull request merged**